lib/npg/samplesheet/novaseq_xseries.pm

```diff
@@ -0,0 +1,905 @@
+  }
+
+
+  my $csv = Text::CSV->new({
+    eol      => qq[\n],
+    sep_char => q[,],
+  });
+
+  my $max_num_columns = max map { scalar @{$_} } @{$self->_all_lines};
+
+  my $file_name = $self->file_name;
+  ## no critic (InputOutput::RequireBriefOpen)
+  open my $fh, q[>], $file_name
+    or croak "Failed to open $file_name for writing";
+
+  for my $line (@{$self->_all_lines}) {
+    my @columns = @{$line};
+    # Pad the row.
+    while (scalar @columns < $max_num_columns) {
+      push @columns, q[];
+    }
+    $csv->print($fh, \@columns);
+  }
+
+  close $fh or carp "Problems closing $file_name";
+  ## use critic
+
+  return;
+}
+
+=head2 add_bclconvert_section
+
+Adds BCLConvert_Settings and BCLConvert_Data sections.
+
+Returns the number of samples added to the section or zero if the section
+has not been added. The latter happens if the number of unique configurations
+for BCLConvert exceeds the maximum number of allowed configurations.
+In this case no further analysis sections should be added to the samplesheet.
+
+=cut
+
+sub add_bclconvert_section {
+  my $self = shift;
+
+  my ($index1_length, $index2_length) = @{$self->_index_reads_length()};
+
+  my @lines = ();
+  push @lines, ['[BCLConvert_Settings]'];
+  push @lines, [q[SoftwareVersion], $SOFTWARE_VERSION];
+
+  # Not clear what CLI analysis option thie coresponds to.
+  # Looks likely to be a list of lanes to run a tag collision check.
+  # According to @srl, bcl-covert tries to correct one error by default
+  # but it checks the tags allow this I.e that they all differ by at least
+  # 3 bases, if they don't it disables the error correction
+  # $add_line->(qw(CombinedIndexCollisionCheck 1;3;4;6));
+
+  # CreateFastqForIndexReads might be an option. Do we need these files?
+  # $add_line->(qw(CreateFastqForIndexReads 1));
+  # When 1 will be appropriate for this trim?
+  # $add_line->(qw(TrimUMI 0));
+  # dragen is the other compression options
+  push @lines, [qw(FastqCompressionFormat gzip)];
+
+  # Barcode mismatch tolerances, the default is 1.
+  # These settings can be omitted.
+  #if ($index1_length) {
+  #  $add_line->(qw(BarcodeMismatchesIndex1 1));
+  #  if ($index2_length) {
+  #    $add_line->(qw(BarcodeMismatchesIndex2 1));
+  #  }
+  #}
+
+  # Adapter trimming settings. The sequence of the Read 1 (or 2) adapter
+  # to be masked or trimmed. To trim multiple adapters, separate the sequences
+  # with a plus sign (+) indicating independent adapters that must be
+  # independently assessed for masking or trimming for each read.
+  # Characters must be A, C, G, or T.
+  # It seems that this settign can also be a column in teh data section
+  #
+  # $add_line->(qw(AdapterRead1 SOME));
+  # $add_line->(qw(AdapterRead2 OTHER));
+
+  push @lines, [];
+  push @lines, ['[BCLConvert_Data]'];
+
+  my @data_header = qw(Lane Sample_ID);
+  if ($index1_length) {
+    push @data_header, q[Index];
+    if ($index2_length) {
```

Comment:
From `_index_reads_length` I understood is possible to get index2_length even if index1_length is not there. Should we be paranoid and check? Guess it may mean a problem. Maybe register issue and add in next version?

Reply:
I am trying not to do too much since we are still finding out what it should be. At the end the code will have to be further refactoted and supplied with proper tests. Then we can decide how paranoid we want to be. Crasy configurations should be rejected by Illumina software.